lib/Core/ExternalDispatcher.cpp

```diff
@@ -39,54 +41,95 @@ void *ExternalDispatcher::resolveSymbol(const std::string &name) {
     if (addr) {
         return addr;
     }
-
     // If it has an asm specifier and starts with an underscore we retry
     // without the underscore. I (DWD) don't know why.
     if (name[0] == 1 && str[0] == '_') {
         ++str;
         addr = llvm::sys::DynamicLibrary::SearchForAddressOfSymbol(str);
     }
 
+    if (!addr) {
+        addr = dlsym(RTLD_DEFAULT, str);
+        if (addr) {
+            llvm::sys::DynamicLibrary::AddSymbol(str, addr);
+        }
+    }
+
     return addr;
 }
 
-bool ExternalDispatcher::call(external_fcn_t targetFunction, const Arguments &args, uint64_t *result,
-                              std::stringstream &err) {
+extern "C" {
+    typedef double (*external_fcn_dd_t)(double);
+    typedef uint32_t (*external_fcn_u32d_t)(double);
+    typedef double (*external_fcn_ddu32_t)(double, uint32_t);
+    typedef double (*external_fcn_ddd_t)(double, double);
+    typedef uint64_t (*external_fcn_t)(...);
+}
+
+uint64_t ExternalDispatcher::double_to_rawbits(double value) {
+    uint64_t bits = 0;
+    memcpy(&bits, &value, 8);
+    return bits;
+}
+
+double ExternalDispatcher::rawbits_to_double(uint64_t bits) {
+    double value = 0.0;
+    memcpy(&value, &bits, 8);
+    return value;
+}
 
-    switch (args.size()) {
-        case 0:
-            *result = targetFunction();
-            break;
-        case 1:
-            *result = targetFunction(args[0]);
-            break;
-        case 2:
-            *result = targetFunction(args[0], args[1]);
-            break;
-        case 3:
-            *result = targetFunction(args[0], args[1], args[2]);
-            break;
-        case 4:
-            *result = targetFunction(args[0], args[1], args[2], args[3]);
-            break;
-        case 5:
-            *result = targetFunction(args[0], args[1], args[2], args[3], args[4]);
-            break;
-        case 6:
-            *result = targetFunction(args[0], args[1], args[2], args[3], args[4], args[5]);
-            break;
-        case 7:
-            *result = targetFunction(args[0], args[1], args[2], args[3], args[4], args[5], args[6]);
-            break;
-        case 8:
-            *result = targetFunction(args[0], args[1], args[2], args[3], args[4], args[5], args[6], args[7]);
-            break;
-        default: {
-            err << "External function has too many parameters";
-            return false;
+bool ExternalDispatcher::call(const std::string& targetName, void *targetAddr, const Arguments &args, uint64_t *result,
+                             std::stringstream &err) {
+    if (targetName == "exp2" || targetName == "log" || targetName == "tan" ||
```

Comment:
A dictionary of name => function pointer would be more efficient

Reply:
Or perhaps look at the argument types instead:
if (retType == double && paramCount==1 && param[0] == double) {
   ...
}